hazelcast/include/hazelcast/client/serialization/pimpl/compact.i.h

```diff
@@ -193,6 +193,33 @@ compact_reader::read()
     return boost::make_optional(std::move(values));
 }
 
+template<typename T>
+typename std::enable_if<std::is_same<std::vector<boost::optional<bool>>,
+                                     typename std::remove_cv<T>::type>::value,
+                        typename boost::optional<T>>::type
+compact_reader::read()
+{
+    int32_t len = object_data_input.read<int32_t>();
+    if (len == 0) {
+        return boost::make_optional<std::vector<boost::optional<bool>>>(
+          std::vector<boost::optional<bool>>(0));
+    }
+    auto values = boost::make_optional<std::vector<boost::optional<bool>>>(
+      std::vector<boost::optional<bool>>(len));
+    int index = 0;
+    byte current_byte = object_data_input.read<byte>();
+    for (int i = 0; i < len; ++i) {
+        if (index == util::Bits::BITS_IN_BYTE) {
+            index = 0;
+            current_byte = object_data_input.read<byte>();
+        }
+        values.value()[i] =
+          boost::make_optional<bool>(((current_byte >> index) & 1) != 0);
```

Comment:
Because we have make_optional here, it is better for readability to introduce the `result` variable here like in java.